CMakeLists.txt

```diff
@@ -22,35 +23,216 @@
 
-if (${CMAKE_SYSTEM_NAME} MATCHES "Windows")
+if (CMAKE_SYSTEM_NAME MATCHES "Linux")
+    add_definitions (-DNN_HAVE_LINUX)
+    find_package (Threads REQUIRED)
+elseif (CMAKE_SYSTEM_NAME MATCHES "Darwin")
+    add_definitions (-DNN_HAVE_OSX)
+    find_package (Threads REQUIRED)
+elseif (CMAKE_SYSTEM_NAME MATCHES "Windows")
+    SET (NN_HAVE_WINSOCK 1)
     add_definitions (-DNN_HAVE_WINDOWS)
     add_definitions (-D_CRT_SECURE_NO_WARNINGS)
+elseif (CMAKE_SYSTEM_NAME MATCHES "FreeBSD")
+    add_definitions (-DNN_HAVE_FREEBSD)
+    find_package (Threads REQUIRED)
+elseif (CMAKE_SYSTEM_NAME MATCHES "NetBSD")
+    add_definitions (-DNN_HAVE_NETBSD)
+    find_package (Threads REQUIRED)
+elseif (CMAKE_SYSTEM_NAME MATCHES "OpenBSD")
+    add_definitions (-DNN_HAVE_OPENBSD)
+    find_package (Threads REQUIRED)
+elseif (CMAKE_SYSTEM_NAME MATCHES "Solaris|SunOS")
+    add_definitions (-DNN_HAVE_SOLARIS)
+    find_package (Threads REQUIRED)
+    check_library_exists (socket socket "" NN_HAVE_SOLARIS_SOCKET)
+    check_library_exists (nsl gethostbyname "" NN_HAVE_SOLARIS_GETHOSTBYNAME)
+elseif (CMAKE_SYSTEM_NAME MATCHES "QNX")
+    add_definitions (-DNN_HAVE_QNX)
+    find_package (Threads REQUIRED)
+    check_library_exists (socket socket "" NN_HAVE_QNX_SOCKET)
+else ()
+    message (AUTHOR_WARNING "WARNING: This platform may or may not be supported: ${CMAKE_SYSTEM_NAME}")
+    message (AUTHOR_WARNING "${ISSUE_REPORT_MSG}")
+endif ()
+
+# Compiler checks.
 
-    if (MINGW)
-        add_definitions (-DNN_HAVE_MINGW -DNN_HAVE_STDINT -D_WIN32_WINNT=0x0600)
-    endif ()
+if (CMAKE_C_COMPILER_ID MATCHES "GNU")
+    add_definitions (-DNN_HAVE_GCC)
+elseif (CMAKE_C_COMPILER_ID MATCHES "Intel")
+    add_definitions (-DNN_HAVE_ICC)
+elseif (CMAKE_C_COMPILER_ID MATCHES "SunPro")
+    add_definitions (-DNN_HAVE_SUN_STUDIO)
+elseif (CMAKE_C_COMPILER_ID MATCHES "Clang")
+    add_definitions (-DNN_HAVE_CLANG)
+elseif (CMAKE_C_COMPILER_ID MATCHES "HP")
+    add_definitions (-DNN_HAVE_HPACC)
+elseif (CMAKE_C_COMPILER_ID MATCHES "MSVC")
+    add_definitions (-DNN_HAVE_MSVC)
+elseif (MINGW)
+    message (FATAL_ERROR "ERROR: MinGW is officially no longer supported.")
 else ()
-    message (FATAL_ERROR "ERROR: CMake build system is intended only to generate MSVC solution files.\nUse autotools build system for any other purpose." )
+    message (AUTHOR_WARNING "WARNING: This compiler may or may not be supported: ${CMAKE_C_COMPILER_ID}")
+    message (AUTHOR_WARNING "${ISSUE_REPORT_MSG}")
 endif ()
 
-#  Build the library itself.
+check_function_exists (gethrtime NN_HAVE_GETHRTIME)
+check_function_exists (socketpair NN_HAVE_SOCKETPAIR)
+check_function_exists (eventfd NN_HAVE_EVENTFD)
+check_function_exists (pipe NN_HAVE_PIPE)
+check_function_exists (pipe2 NN_HAVE_PIPE2)
+check_function_exists (accept4 NN_HAVE_ACCEPT4)
+check_function_exists (epoll_create NN_HAVE_EPOLL)
+check_function_exists (kqueue NN_HAVE_KQUEUE)
+check_function_exists (poll NN_HAVE_POLL)
```

Comment:
A lot of this is ported directly from autotools, and could be stale. Let's try to get CMake mainlined today -- even along with cruft that is ported wholesale from autotools -- and then continue with the fine-tooth comb once mainlined.

Good strategy?